ocaml/xapi/xapi_sm.ml

```diff
@@ -92,22 +93,29 @@ let unregister_plugin ~__context query_result =
 	let driver = String.lowercase query_result.driver in
 	if is_v1 query_result.required_api_version then begin
 		info "Not unregistering SM plugin %s (required_api_version %s < 2.0)" driver query_result.required_api_version;
-	end else begin
-		let existing = List.map (fun (rf, rc) -> rc.API.sM_type, (rf, rc)) (Db.SM.get_all_records ~__context) in
-		if List.mem_assoc driver existing then begin
-			info "Unregistering SM plugin %s (version %s)" driver query_result.version;
-			try
-				Db.SM.destroy ~__context ~self:(fst (List.assoc driver existing))
-			with _ -> ()
-		end
-	end
+	end else
+		List.iter
+			(fun (rf, rc) ->
+				if rc.API.sM_type = driver then
+					try
+						info "Unregistering SM plugin %s (version %s)" driver query_result.version;
+						Db.SM.destroy ~__context ~self:rf
+					with e ->
+						warn "Ignore unregistering SM plugin failure: %s" (Printexc.to_string e))
+			(Db.SM.get_all_records ~__context)
+
+(* Lock to serialize SM plugin registration. Not necessary for unregister_plugin,
+   which is never called externally but only as a step in register_plugin. *)
```

Comment:
At present `unregister_plugin` is not called externally from anywhere, but I feel we should ensure that this remains true in future. To do that we could have `xapi_sm.mli` without that function, or we could remove the function and move its `List.iter...` into `register_plugin`, replacing the call to `unregister_plugin`.